### PR TITLE
Fix: Migrate to WHATWG URL standard

### PR DESCRIPTION
**Pull request by sarvaje**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [x] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/getting-started/pull-requests/#commit-messagess)

For non-trivial changes, please make sure you also:

- [ ] Added/Updated related documentation.
- [x] Added/Updated related tests.

### CODE REVIEW

**Comment by sarvaje** [comment]
@molant @alrra this PR is ready to review

File: packages/rule-apple-touch-icons/src/rule.ts [hunk 1]
```diff
@@ -99,18 +99,12 @@ export default class AppleTouchIconsRule implements IRule {
...
              * If `href` exists and is not an empty string, try
              * to figure out the full URL of the `apple-touch-icon`.
              */
 
-            if (url.parse(appleTouchIconHref).protocol) {
-                appleTouchIconURL = appleTouchIconHref;
-            } else {
-                appleTouchIconURL = url.resolve(resource, appleTouchIconHref);
-            }
+            const appleTouchIconURL = new URL(appleTouchIconHref, resource).href;
```
Comment (alrra) [inline packages/rule-apple-touch-icons/src/rule.ts]: This is great!

File: packages/sonarwhal/src/lib/utils/misc.ts [hunk 2]
```diff
@@ -128,20 +136,19 @@ const normalizeString = (value: string, defaultValue?: string): string => {
     return value.toLowerCase().trim();
 };
 
+const protocolRegex = /([^:]*):.*/;
```
Comment (alrra) [inline packages/sonarwhal/src/lib/utils/misc.ts]: This also matches `:test`, shouldn't it be `/([^:]+):.*/`?

File: packages/sonarwhal/src/lib/utils/misc.ts [hunk 3]
```diff
@@ -49,12 +49,20 @@ const getFileName = (resource: string) => {
...
-     * The url needs to be parsed first
-     * otherwise the result from path.extname could be incorrect, e.g.: https://sonarwhal.com => '.com'
-     */
-    return path.extname(url.parse(resource).pathname)
-        .split('.')
+    let url: URL;
+
+    try {
+        /*
+         * he url needs to be parsed first
```
Comment (alrra) [inline packages/sonarwhal/src/lib/utils/misc.ts]: > **T**he

File: packages/sonarwhal/src/lib/utils/misc.ts (hunk 2 above)
Comment (sarvaje) [reply packages/sonarwhal/src/lib/utils/misc.ts]: `isRegularProtocol(':test')` should return `true` or `false`?

Comment (sarvaje) [reply packages/sonarwhal/src/lib/utils/misc.ts]: With the current regex it will return `false` because the protocol is the empty string.

Comment (alrra) [reply packages/sonarwhal/src/lib/utils/misc.ts]: IMHO `:test` shouldn't match.

Comment (alrra) [reply packages/sonarwhal/src/lib/utils/misc.ts]: Yes, nevermind, let's just readd [this comment](https://github.com/sonarwhal/sonarwhal/pull/903/files#diff-99cb63c80e7850a78eac3f222eac6898L139) back.

File: packages/sonarwhal/src/lib/sonarwhal.ts [hunk 4]
```diff
@@ -237,11 +237,11 @@ export class Sonarwhal extends EventEmitter {
     }
 
     /** Runs all the configured rules and plugins on a target */
-    public async executeOn(target: url.Url): Promise<Array<Problem>> {
+    public async executeOn(target: url.URL): Promise<Array<Problem>> {
 
         const start: number = Date.now();
 
-        debug(`Starting the analysis on ${target.path}`);
+        debug(`Starting the analysis on ${target.pathname}${target.search}`);
```
[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/sonarwhal/src/lib/sonarwhal.ts]: sure thing!

[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/sonarwhal/src/lib/sonarwhal.ts]: @molant done!

**Comment by alrra** [comment]
FYI: I'm going to change this to `Chore:` since this is just refactoring.

**Pull request closed without merging**